composer/trainer/trainer.py

```diff
@@ -997,6 +993,79 @@ def _train_loop(self) -> None:
                                                                                    event=Event.EPOCH_END):
                 self._checkpoint_saver.save_checkpoint(state=self.state, seed=self._seed, device=self._device)
 
+    def _handle_cuda_oom(self, e: RuntimeError):
+        """Handles CUDA Out of Memory and rescales if using adaptive grad_accum."""
+        if self.adaptive_grad_accum and "CUDA out of memory" in str(e):
+            # Raise runtime error if training 1 sample at a time still resulted in CUDA out of memory
+            if self.state.grad_accum == self.state.batch_num_samples:
+                raise RuntimeError("CUDA out of memory. Train loop failed with an internal microbatch of size 1")
+            else:
+                self.state.grad_accum = min(2 * self.state.grad_accum, self.state.batch_num_samples)
+                self.logger.metric_batch({'trainer/grad_accum': self.state.grad_accum})
+        else:
+            # If not CUDA out of memory, raise exception to user. Note that this
+            # truncates the call stack back only to this newly raised error
+            raise e
+
+    def _compute_metrics(self, train_metrics: MetricCollection):
+        """Compute training metrics. Adaptively change microbatch size if enabled to maximize GPU usage.
+
+        Args:
+            train_metrics (MetricCollection): Existing train metrics
+        """
+        rerun = True
+        while rerun:
+            try:
+                self.state.model.eval()
+                with torch.no_grad():
+                    # Store results so we only update train_metrics if all batches are processed without OOM
+                    results = []
+                    for eval_microbatch in self._train_data_spec.split_batch(self.state.batch, self.state.grad_accum):
+                        # TODO: Detect if self.run_event(Event.AFTER_DATALOADER) changes the training
+                        # data and if so print a warning that metrics may return unexpected results
+                        outputs, targets = self._original_model.validate(eval_microbatch)
+                        results.append((outputs, targets))
+                    for outputs, targets in results:
+                        train_metrics.update(outputs, targets)
+                rerun = False
+            except RuntimeError as e:
+                self._handle_cuda_oom(e)
```

Comment:
We can get by without wrapping this for the time being. It's very unlikely an OOM would be caused under the `torch.no_grad` context.

Reply:
This happens before `_train_and_compute_loss()`, so if enabled, I think we'd see OOM here on the very first pass. I agree it's not the bottleneck step because of `torch.no_grad`, but I think it's required because it happens first

Reply:
Talked with @ajaysaini725, and I think there's a number of reasons why we shouldn't wrap this:

Since the memory requirements of this step are lower than that of an ordinary training fwd/bwd pass, it doesn't actually make sense to use `state.grad_accum` at all to chunk the eval microbatches here. The optimal grad_accum for use here is different than what should be used for the main training loop.

Because of this, suppose an initial OOM in this step leads to some grad accum tuning to make this function succeed. Then suppose that on the first full train batch size, the grad accum is further increased so that _that_ can pass. That means that when this function next runs, the grad accum will have changed, and it won't need as much memory as it did before. That might sound like a good thing, but it can cause memory fragmentation issues. Furthermore, we'll be running this function less efficiently than we otherwise could (and computing train metrics can sometimes impose as much as a 20% slowdown, so that's a big deal).

Anyways, @ajaysaini725 had the insight that the microbatch size we should use, in theory, should be exactly the same as the per-GPU batch size we use for regular eval. So in theory, the wrapper for this should really be a part of our eventual eval_batch_size tuning, not our grad_accum tuning.

In summary, I think there's a lot of complexity going on here that would best be tackled in a separate PR. Ultimately, that PR isn't urgent, since training metrics just aren't used that frequently by us anyways.

Reply:
Got it, I'm convinced :) 

Will gut the `while try except` part but keep the function decomposition